### PR TITLE
update Middleware

### PR DESCRIPTION
**Pull request by Rick-Anderson**

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@blowdart  can you approve the security changes I've made to this doc?

**Comment by Rick-Anderson** [comment]
@ardalis  @danroth27 please review changes

**Comment by blowdart** [comment]
Works for me

File: aspnet/fundamentals/middleware.rst [hunk 1]
```diff
@@ -28,52 +28,61 @@ The ASP.NET request pipeline consists of a sequence of request delegates, called
...
 
 .. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
 	:language: c#
-	:linenos:
 	:lines: 20-31
 	:emphasize-lines: 5
 	:dedent: 8
 
-You chain multiple request delegates together making a different call, with a ``next`` parameter representing the next delegate in the pipeline. Note that just because you're calling it "next" doesn't mean you can't perform actions both before and after the next delegate, as this example demonstrates:
+You chain multiple request delegates together; the ``next`` parameter represents the next delegate in the pipeline. You can typically perform actions both before and after the next delegate, as this example demonstrates:
```
Comment (ardalis) [inline aspnet/fundamentals/middleware.rst]: Note either here or above where you reference App.Run that you can also terminate the pipeline by not calling the ``next`` parameter.

**Comment by ardalis** [comment]
One minor fix then :shipit:

**Comment by rachelappel** [comment]
merged 80dfdcbc0babaedcde50991d2b8409133001712c

File: aspnet/fundamentals/middleware.rst [hunk 2]
```diff
@@ -2,24 +2,24 @@
 
 Middleware
 ==========
-By `Steve Smith`_
+By `Steve Smith`_ and `Rick Anderson`_
 
-Small application components that can be incorporated into an HTTP request pipeline are known collectively as middleware. ASP.NET 5 has integrated support for middleware, which are wired up in an application's ``Configure`` method during :doc:`startup`.
+Middleware is defined in the `OWIN specification <http://owin.org/html/spec/owin-1.0.1.html>`__ as pass through components that form a pipeline between a server and application to inspect, route, or modify request and response messages for a specific purpose. Middleware consists of application components are incorporated into the ASP.NET HTTP pipeline.
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: > Middleware is defined in the `` `OWIN specification <http://owin.org/html/spec/owin-1.0.1.html>`__  ``

We should remove this reference to OWIN. ASP.NET Core middleware and OWIN middleware are *not* the same

Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: > the ASP.NET HTTP pipeline

the ASP.NET Core HTTP pipeline

File: aspnet/fundamentals/middleware.rst [hunk 3]
```diff
@@ -2,24 +2,24 @@
 
 Middleware
 ==========
-By `Steve Smith`_
+By `Steve Smith`_ and `Rick Anderson`_
 
-Small application components that can be incorporated into an HTTP request pipeline are known collectively as middleware. ASP.NET 5 has integrated support for middleware, which are wired up in an application's ``Configure`` method during :doc:`startup`.
+Middleware is defined in the `OWIN specification <http://owin.org/html/spec/owin-1.0.1.html>`__ as pass through components that form a pipeline between a server and application to inspect, route, or modify request and response messages for a specific purpose. Middleware consists of application components are incorporated into the ASP.NET HTTP pipeline. 
 
 In this article:
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: Replace this bulleted list with the `contents` directive.

File: aspnet/fundamentals/middleware.rst [hunk 4]
```diff
@@ -177,15 +197,13 @@ Testing the middleware (by setting the ``Hosting:Environment`` environment varia
...
-
-Summary
--------
-
-Middleware provide simple components for adding features to individual web requests. Applications configure their request pipelines in accordance with the features they need to support, and thus have fine-grained control over the functionality each request uses. Developers can easily create their own middleware to provide additional functionality to ASP.NET applications.
+.. note:: The `UseStaticFiles <https://github.com/aspnet/StaticFiles/blob/1.0.0-beta6/src/Microsoft.AspNet.StaticFiles/StaticFileExtensions.cs#L44>`_ extension method (which creates the `StaticFileMiddleware <https://github.com/aspnet/StaticFiles/blob/1.0.0-beta6/src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs#L30>`_) also uses ``UseMiddleware<T>``. In this case, the ``StaticFileOptions`` parameter is passed in, but other constructor parameters are supplied by ``UseMiddleware<T>`` and dependency injection.
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: These beta6 GitHub links should be removed. We should remove all links to product code on GitHub as they are too brittle to maintain.

File: aspnet/fundamentals/middleware.rst [hunk 5]
```diff
@@ -144,20 +165,20 @@ ASP.NET ships with the following middleware components:
...
 
 .. literalinclude:: middleware/sample/src/MiddlewareSample/RequestLoggerMiddleware.cs
 	:language: c#
 	:caption: RequestLoggerMiddleware.cs
-	:linenos:
 	:emphasize-lines: 13, 19
 
 The middleware follows the `Explicit Dependencies Principle <http://deviq.com/explicit-dependencies-principle/>`_ and exposes all of its dependencies in its constructor. Middleware can take advantage of the `UseMiddleware<T> <https://github.com/aspnet/HttpAbstractions/blob/1.0.0-beta6/src/Microsoft.AspNet.Http.Abstractions/Extensions/UseMiddlewareExtensions.cs>`_ extension to inject services directly into their constructors, as shown in the example below. Dependency injected services are automatically filled, and the extension takes a ``params`` array of arguments to be used for non-injected parameters.
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: These beta6 GitHub links should be removed. We should remove all links to product code on GitHub as they are too brittle to maintain.

File: aspnet/fundamentals/middleware.rst [hunk 6]
```diff
@@ -144,20 +165,20 @@ ASP.NET ships with the following middleware components:
 Writing middleware
 ------------------
 
+The `CodeLabs middleware tutorial <https://github.com/Microsoft-Build-2016/CodeLabs-WebDev/tree/master/Module2-AspNetCore>`__ provides a good introduction to writing middleware.
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: Let's put this in the related content section. These docs should be as stand-alone as possible.

File: aspnet/fundamentals/middleware.rst [hunk 7]
```diff
@@ -28,52 +28,61 @@ The ASP.NET request pipeline consists of a sequence of request delegates, called
...
 
 .. literalinclude:: middleware/sample/src/MiddlewareSample/Startup.cs
 	:language: c#
-	:linenos:
 	:lines: 34-49
 	:emphasize-lines: 5,8,14
 	:dedent: 8
 
-.. warning:: Be wary of modifying ``HttpResponse`` after invoking next, since one of the components further down the pipeline may have written to the response, causing it to be sent to the client.
+.. warning:: Avoid modifying ``HttpResponse`` after invoking next, one of the next components in the pipeline may have written to the response, causing it to be sent to the client.
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: > Avoid modifying ``HttpResponse`` after invoking next, one of the next components in the pipeline may have written to the response, causing it to be sent to the client.  

Avoid modifying ``HttpResponse`` after invoking ``next``. One of the next components in the pipeline may have written to the response causing it to be sent to the client.

File: aspnet/fundamentals/middleware.rst [hunk 8]
```diff
@@ -2,24 +2,24 @@
...
-Small application components that can be incorporated into an HTTP request pipeline are known collectively as middleware. ASP.NET 5 has integrated support for middleware, which are wired up in an application's ``Configure`` method during :doc:`startup`.
+Middleware is defined in the `OWIN specification <http://owin.org/html/spec/owin-1.0.1.html>`__ as pass through components that form a pipeline between a server and application to inspect, route, or modify request and response messages for a specific purpose. Middleware consists of application components are incorporated into the ASP.NET HTTP pipeline. 
 
 In this article:
 	- `What is middleware`_
 	- `Creating a middleware pipeline with IApplicationBuilder`_
 	- `Built-in middleware`_
 	- `Writing middleware`_
 
-`Download sample from GitHub <https://github.com/aspnet/Docs/tree/master/aspnet/fundamentals/middleware/sample>`_.
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: Not sure why this was removed - we should still provide a sample link.

File: aspnet/fundamentals/middleware.rst [hunk 9]
```diff
@@ -2,24 +2,24 @@
...
+
+Request delegates are configured using `Run <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/RunExtensions/index.html>`__, `Map <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/MapExtensions/index.html?highlight=microsoft.aspnet.builder.mapextensions#Microsoft.AspNet.Builder.MapExtensions.Map>`__, and `Use <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/UseExtensions/index.html?highlight=microsoft.aspnet.builder.useextensions#Microsoft.AspNet.Builder.UseExtensions.Use>`__ extension methods on the `IApplicationBuilder <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/IApplicationBuilder/index.html>`_ type that is passed into the ``Configure`` method in the ``Startup`` class. An individual request delegate can be specified in-line as an anonymous method, or it can be defined in a reusable class. These reusable classes  are `middleware`, or `middleware components`. Each middleware component in the request pipeline is responsible for invoking the next component in the pipeline, or short-circuiting the chain if appropriate.
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: > https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/UseExtensions/index.html?highlight=microsoft.aspnet.builder.useextensions#Microsoft.AspNet.Builder.UseExtensions.Use

Remove the highlight query string parameter

File: aspnet/fundamentals/middleware.rst [hunk 10]
```diff
@@ -82,11 +91,10 @@ In the above example, the call to ``await next.Invoke()`` will call into the del
...
+You configure the HTTP pipeline using `Run <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/RunExtensions/index.html>`__, `Map <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/MapExtensions/index.html?highlight=microsoft.aspnet.builder.mapextensions#Microsoft.AspNet.Builder.MapExtensions.Map>`__,  and `Use <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/UseExtensions/index.html?highlight=microsoft.aspnet.builder.useextensions#Microsoft.AspNet.Builder.UseExtensions.Use>`__. The ``Run`` method short circuits the pipeline (that is, it will not call a ``next`` request delegate). Thus, ``Run`` should only be called at the end of your pipeline. ``Run`` is a convention, and some middleware components may expose their own Run[Middleware] methods that should only run at the end of the pipeline. The following two middleware are equivalent as the ``Use`` version doesn't use the ``next`` parameter:
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: Remove the highlight query string parameter from the URIs

File: aspnet/fundamentals/middleware.rst [hunk 11]
```diff
@@ -28,52 +28,61 @@ The ASP.NET request pipeline consists of a sequence of request delegates, called
...
+
+You can see an example of setting up the request pipeline in the default web site template that ships with Visual Studio 2015. The ``Configure`` method adds the following middleware components:
 
-You can see an example of setting up a request pipeline, using a variety of request delegates, in the default web site template that ships with Visual Studio 2015. Its ``Configure`` method, shown below, first wires up error pages (in development) or the site's production error handler, then builds out the pipeline with support for static files, ASP.NET Identity authentication, and finally, ASP.NET MVC.
+#. Error handling (for both development and non-development environments)
+#. IIS HttpPlatformHandler reverse proxy module. This module handles forwarded Windows Authentication, request schemes, remote IPs, and so on.
+#. Static file server
+#. ASP.NET Identity authentication
+#. ASP.NET MVC
```
Comment (danroth27) [inline aspnet/fundamentals/middleware.rst]: > ASP.NET Identity authentication
> ASP.NET MVC

These are the names of the existing frameworks. The new ones are called ASP.NET Core Identity and ASP.NET Core MVC (or just Identity and MVC if ASP.NET Core is obvious from the context).

**Pull request closed without merging**